File: tetris.py
# tetris.py
import curses
import random
import time

# Tetromino shapes
SHAPES = {
    'I': [[1,1,1,1]],
    'O': [[1,1],[1,1]],
    'T': [[0,1,0],[1,1,1]],
    'S': [[0,1,1],[1,1,0]],
    'Z': [[1,1,0],[0,1,1]],
    'J': [[1,0,0],[1,1,1]],
    'L': [[0,0,1],[1,1,1]]
}
COLORS = {'I': 1, 'O': 2, 'T': 3, 'S': 4, 'Z': 5, 'J': 6, 'L': 7}

class Tetris:
    def __init__(self, stdscr):
        self.stdscr = stdscr
        self.width = 10
        self.height = 20
        self.board = [[0]*self.width for _ in range(self.height)]
        self.score = 0
        self.lines = 0
        self.level = 1
        self.game_over = False
        self.paused = False
        self.fall_time = 0
        self.fall_speed = 0.5  # seconds per drop
        self.current_piece = None
        self.next_piece = None
        self.bag = []
        self.init_colors()
        self.spawn_piece()
        self.next_piece = self.get_piece()

    def init_colors(self):
        curses.start_color()
        curses.init_pair(1, curses.COLOR_CYAN, curses.COLOR_BLACK)
        curses.init_pair(2, curses.COLOR_YELLOW, curses.COLOR_BLACK)
        curses.init_pair(3, curses.COLOR_MAGENTA, curses.COLOR_BLACK)
        curses.init_pair(4, curses.COLOR_GREEN, curses.COLOR_BLACK)
        curses.init_pair(5, curses.COLOR_RED, curses.COLOR_BLACK)
        curses.init_pair(6, curses.COLOR_BLUE, curses.COLOR_BLACK)
        curses.init_pair(7, curses.COLOR_WHITE, curses.COLOR_BLACK)

    def get_piece(self):
        if not self.bag:
            self.bag = list(SHAPES.keys())
            random.shuffle(self.bag)
        shape = self.bag.pop()
        return {'shape': shape, 'matrix': SHAPES[shape], 'x': self.width//2 - len(SHAPES[shape][0])//2, 'y': 0, 'color': COLORS[shape]}

    def spawn_piece(self):
        self.current_piece = self.next_piece or self.get_piece()
        self.next_piece = self.get_piece()
        if self.check_collision(self.current_piece['matrix'], self.current_piece['x'], self.current_piece['y']):
            self.game_over = True

    def check_collision(self, matrix, x, y):
        for row_idx, row in enumerate(matrix):
            for col_idx, cell in enumerate(row):
                if cell:
                    board_x = x + col_idx
                    board_y = y + row_idx
                    if board_x < 0 or board_x >= self.width or board_y >= self.height or board_y < 0:
                        return True
                    if board_y >= 0 and self.board[board_y][board_x]:
                        return True
        return False

    def lock_piece(self):
        piece = self.current_piece
        for row_idx, row in enumerate(piece['matrix']):
            for col_idx, cell in enumerate(row):
                if cell:
                    board_x = piece['x'] + col_idx
                    board_y = piece['y'] + row_idx
                    if board_y >= 0:
                        self.board[board_y][board_x] = piece['color']
        self.clear_lines()
        self.spawn_piece()

    def clear_lines(self):
        lines_cleared = 0
        for y in range(self.height-1, -1, -1):
            if all(self.board[y]):
                del self.board[y]
                self.board.insert(0, [0]*self.width)
                lines_cleared += 1
        if lines_cleared:
            self.lines += lines_cleared
            self.score += lines_cleared * 100
            self.level = self.lines // 10 + 1
            self.fall_speed = max(0.1, 0.5 - (self.level-1) * 0.04)

    def rotate(self):
        piece = self.current_piece
        matrix = piece['matrix']
        rotated = [list(row) for row in zip(*matrix[::-1])]  # clockwise
        if not self.check_collision(rotated, piece['x'], piece['y']):
            piece['matrix'] = rotated

    def move(self, dx, dy):
        piece = self.current_piece
        if not self.check_collision(piece['matrix'], piece['x']+dx, piece['y']+dy):
            piece['x'] += dx
            piece['y'] += dy
            return True
        elif dy == 1:  # hard drop or landing
            self.lock_piece()
        return False

    def hard_drop(self):
        while not self.move(0, 1):
            pass

    def update(self):
        if self.paused or self.game_over:
            return
        self.fall_time += 0.016  # approx 60 FPS
        if self.fall_time >= self.fall_speed:
            self.fall_time = 0
            if not self.move(0, 1):
                self.lock_piece()

    def draw(self):
        self.stdscr.clear()
        h, w = self.stdscr.getmaxyx()
        # Draw board
        for y in range(self.height):
            for x in range(self.width):
                color = self.board[y][x]
                if color:
                    self.stdscr.addch(y+1, x*2+2, '  ', curses.color_pair(color))
                else:
                    self.stdscr.addch(y+1, x*2+2, '..')
        # Draw current piece
        if self.current_piece:
            piece = self.current_piece
            for row_idx, row in enumerate(piece['matrix']):
                for col_idx, cell in enumerate(row):
                    if cell:
                        board_x = (piece['x'] + col_idx) * 2 + 2
                        board_y = piece['y'] + row_idx + 1
                        if 0 <= board_y < self.height+1 and 0 <= board_x < self.width*2+2:
                            self.stdscr.addch(board_y, board_x, '  ', curses.color_pair(piece['color']))
        # Draw info
        self.stdscr.addstr(0, 2, f"Score: {self.score}  Lines: {self.lines}  Level: {self.level}")
        # Next piece
        if self.next_piece:
            self.stdscr.addstr(1, self.width*2+5, "Next:")
            for row_idx, row in enumerate(self.next_piece['matrix']):
                for col_idx, cell in enumerate(row):
                    if cell:
                        self.stdscr.addch(2+row_idx, self.width*2+7+col_idx*2, '  ', curses.color_pair(self.next_piece['color']))
        # Controls
        self.stdscr.addstr(self.height+2, 2, "Controls: ←/→ move, ↑ rotate, ↓ soft drop, Space hard drop, P pause, Q quit")
        if self.game_over:
            self.stdscr.addstr(self.height+4, 2, "GAME OVER! Press Q to quit.")
        elif self.paused:
            self.stdscr.addstr(self.height+4, 2, "PAUSED. Press P to resume.")
        self.stdscr.refresh()

    def run(self):
        self.stdscr.nodelay(1)
        self.stdscr.timeout(16)  # ~60 FPS
        while not self.game_over:
            key = self.stdscr.getch()
            if key == ord('q') or key == ord('Q'):
                break
            elif key == ord('p') or key == ord('P'):
                self.paused = not self.paused
            if not self.paused and not self.game_over:
                if key == curses.KEY_LEFT or key == ord('a') or key == ord('A'):
                    self.move(-1, 0)
                elif key == curses.KEY_RIGHT or key == ord('d') or key == ord('D'):
                    self.move(1, 0)
                elif key == curses.KEY_DOWN or key == ord('s') or key == ord('S'):
                    self.move(0, 1)
                elif key == curses.KEY_UP or key == ord('w') or key == ord('W'):
                    self.rotate()
                elif key == ord(' '):
                    self.hard_drop()
            self.update()
            self.draw()
        curses.endwin()

def main(stdscr):
    game = Tetris(stdscr)
    game.run()

if __name__ == "__main__":
    curses.wrapper(main)

File: tetris.js
// tetris.js
const readline = require('readline');
const { createInterface } = require('readline');

const W = 10, H = 20;
const SHAPES = {
  I: [[1,1,1,1]],
  O: [[1,1],[1,1]],
  T: [[0,1,0],[1,1,1]],
  S: [[0,1,1],[1,1,0]],
  Z: [[1,1,0],[0,1,1]],
  J: [[1,0,0],[1,1,1]],
  L: [[0,0,1],[1,1,1]]
};
const SHAPE_NAMES = ['I','O','T','S','Z','J','L'];

class Tetris {
  constructor() {
    this.board = Array.from({length: H}, () => Array(W).fill(0));
    this.score = 0;
    this.lines = 0;
    this.level = 1;
    this.fallSpeed = 500; // ms
    this.gameOver = false;
    this.paused = false;
    this.bag = [];
    this.current = this.getPiece();
    this.next = this.getPiece();
    this.lastDrop = Date.now();
    this.rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    readline.emitKeypressEvents(process.stdin);
    process.stdin.setRawMode(true);
    process.stdin.on('keypress', (str, key) => this.handleKey(str, key));
  }

  getPiece() {
    if (this.bag.length === 0) {
      this.bag = [...SHAPE_NAMES];
      for (let i = this.bag.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [this.bag[i], this.bag[j]] = [this.bag[j], this.bag[i]];
      }
    }
    const name = this.bag.pop();
    const matrix = SHAPES[name].map(row => [...row]);
    const color = Math.floor(Math.random() * 7) + 1;
    return { name, matrix, x: Math.floor((W - matrix[0].length) / 2), y: 0, color };
  }

  checkCollision(matrix, x, y) {
    for (let row = 0; row < matrix.length; row++) {
      for (let col = 0; col < matrix[0].length; col++) {
        if (matrix[row][col]) {
          const bx = x + col;
          const by = y + row;
          if (bx < 0 || bx >= W || by >= H || by < 0) return true;
          if (by >= 0 && this.board[by][bx]) return true;
        }
      }
    }
    return false;
  }

  lockPiece() {
    const p = this.current;
    for (let row = 0; row < p.matrix.length; row++) {
      for (let col = 0; col < p.matrix[0].length; col++) {
        if (p.matrix[row][col]) {
          const bx = p.x + col;
          const by = p.y + row;
          if (by >= 0) this.board[by][bx] = p.color;
        }
      }
    }
    this.clearLines();
    this.current = this.next;
    this.next = this.getPiece();
    if (this.checkCollision(this.current.matrix, this.current.x, this.current.y)) {
      this.gameOver = true;
    }
  }

  clearLines() {
    let cleared = 0;
    for (let y = H - 1; y >= 0; y--) {
      if (this.board[y].every(cell => cell !== 0)) {
        this.board.splice(y, 1);
        this.board.unshift(Array(W).fill(0));
        cleared++;
        y++;
      }
    }
    if (cleared) {
      this.lines += cleared;
      this.score += cleared * 100;
      this.level = Math.floor(this.lines / 10) + 1;
      this.fallSpeed = Math.max(100, 500 - (this.level - 1) * 40);
    }
  }

  rotate() {
    const p = this.current;
    const rotated = p.matrix[0].map((_, idx) => p.matrix.map(row => row[idx]).reverse());
    if (!this.checkCollision(rotated, p.x, p.y)) {
      p.matrix = rotated;
    }
  }

  move(dx, dy) {
    const p = this.current;
    if (!this.checkCollision(p.matrix, p.x + dx, p.y + dy)) {
      p.x += dx;
      p.y += dy;
      return true;
    }
    if (dy === 1) this.lockPiece();
    return false;
  }

  hardDrop() {
    while (this.move(0, 1)) {}
  }

  update() {
    if (this.paused || this.gameOver) return;
    const now = Date.now();
    if (now - this.lastDrop >= this.fallSpeed) {
      this.lastDrop = now;
      if (!this.move(0, 1)) this.lockPiece();
    }
  }

  draw() {
    console.clear();
    console.log(`Score: ${this.score}  Lines: ${this.lines}  Level: ${this.level}`);
    for (let y = 0; y < H; y++) {
      let row = '|';
      for (let x = 0; x < W; x++) {
        row += this.board[y][x] ? '[]' : '  ';
      }
      row += '|';
      console.log(row);
    }
    // Draw current piece (overlay)
    if (this.current) {
      const p = this.current;
      for (let row = 0; row < p.matrix.length; row++) {
        for (let col = 0; col < p.matrix[0].length; col++) {
          if (p.matrix[row][col]) {
            const bx = p.x + col;
            const by = p.y + row;
            // Move cursor and print
            process.stdout.write(`\x1b[${by+3};${bx*2+2}H`);
            process.stdout.write('[]');
          }
        }
      }
    }
    // Next piece
    console.log(`\x1b[3;${W*2+6}HNext:`);
    if (this.next) {
      for (let row = 0; row < this.next.matrix.length; row++) {
        for (let col = 0; col < this.next.matrix[0].length; col++) {
          if (this.next.matrix[row][col]) {
            process.stdout.write(`\x1b[${4+row};${W*2+8+col*2}H`);
            process.stdout.write('[]');
          }
        }
      }
    }
    console.log(`\x1b[${H+4};0HControls: ←/→ move, ↑ rotate, ↓ soft drop, Space hard drop, P pause, Q quit`);
    if (this.gameOver) console.log(`\x1b[${H+6};0HGAME OVER! Press Q to quit.`);
    else if (this.paused) console.log(`\x1b[${H+6};0HPAUSED. Press P to resume.`);
  }

  handleKey(str, key) {
    if (key.ctrl && key.name === 'c') process.exit();
    if (key.name === 'q') process.exit();
    if (key.name === 'p') { this.paused = !this.paused; return; }
    if (this.paused || this.gameOver) return;
    const name = key.name;
    if (name === 'left') this.move(-1, 0);
    else if (name === 'right') this.move(1, 0);
    else if (name === 'down') this.move(0, 1);
    else if (name === 'up') this.rotate();
    else if (name === 'space') this.hardDrop();
  }

  run() {
    this.lastDrop = Date.now();
    const interval = setInterval(() => {
      this.update();
      this.draw();
      if (this.gameOver) clearInterval(interval);
    }, 16);
  }
}

const game = new Tetris();
game.run();
